src/txmempool.h

```diff
@@ -538,6 +538,29 @@ class CTxMemPool
     CTxMemPool(const CFeeRate &_minReasonableRelayFee);
     ~CTxMemPool();
 
+    /** Atomically (with respect to the mempool) call f on each mempool entry, and then clear the mempool */
+    template <typename Lambda>
+    void forEachThenClear(const Lambda &f)
+    {
+        WRITELOCK(cs);
+        for (CTxMemPool::indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++)
+        {
+            f(*it);
+        }
+        _clear();
+    }
+
+    /** Atomically (with respect to the mempool) call f on each mempool entry */
+    template <typename Lambda>
+    void forEach(const Lambda &f)
+    {
+        WRITELOCK(cs);
+        for (CTxMemPool::indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++)
+        {
+            f(*it);
+        }
```

Comment:
Also need to resubmit and clear the txCommitQ.

Reply:
done, but in the caller of this function.